### PR TITLE
feat: check that init of base class has been called

### PR DESCRIPTION
**Pull request by domire8**
## Description

This is something that has bothered me a little bit for a while. If you check the modulo_controllers::ControllerInterface, you can see that all `on_XXX` transition callbacks are overridden, except the `on_init`. We chose to override that to assure that the base class `on_XXX` steps are always called, i.e. to not put the responsibility of putting `return ControllerInterface::on_XXX` into the derived class' callback onto the user.

However, this doesn't work with the `on_init` callback since it already has no arguments in the ros2 control base class. Meaning we now have mixed behavior where a derived class has to call the `on_init` of the parent explicitly, but not the other callbacks. One way that I can think of to remedy that is to check that the `on_init` has been called with a flag in the base class, as proposed here.

If you can see better solutions, I'd be happy to hear them.

## Review guidelines

Estimated Time of Review: 5 minutes

## Checklist before merging:

- [ ] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

File: source/modulo_controllers/src/ControllerInterface.cpp [hunk 1]
```diff
@@ -52,6 +54,12 @@ rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn Contro
 
 rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
 ControllerInterface::on_configure(const rclcpp_lifecycle::State&) {
+  if (!on_init_called_) {
+    RCLCPP_ERROR(
+        get_node()->get_logger(),
+        "'ModuloControllerInterface::on_init()' has not been called, any derived class needs to do that");
```
Comment (eeberhard) [inline source/modulo_controllers/src/ControllerInterface.cpp:60]: If we go with this approach, I would reword it:
```suggestion
        "The controller has not been properly initialized! Derived controller classes must call 'ControllerInterface::on_init()' during their own initialization before being configured.");
```

Also the class is now called `ControllerInterface`, not `ModuloControllerInterface`

**Review by eeberhard — COMMENTED** [review]
An alternative is to mark `ControllerInterface::on_init() final`, and then define a new specific `virtual ControllerInterface::add_interfaces()`. The intention f

```c++
ControllerInterface::on_init() {
  parameter_cb_handle_ = get_node()->add_on_set_parameters_callback(...);
  try {
    // add base parameters
    add_parameter(...);
    //...

    // add any parameters, signals, services or predicates of the derived class
    return this->add_interfaces();
  } catch (...) {
    RCLCPP_ERROR(...);
  }
  return CallbackReturn::ERROR;
}
```

This means that derived controllers no longer have to do:
```c++
AdmittanceController::on_init() {
  auto ret = RobotControllerInterface::on_init();
  if (ret != CallbackReturn::SUCCESS) {
    return ret;
  }

  try {
    add_parameter(max_linear_velocity_, "Maximum linear velocity to clamp the twist command in meters per second");
    //...
    add_input<CartesianWrench>("ft_sensor");
  } catch (const std::exception& e) {
    RCLCPP_ERROR(get_node()->get_logger(), "Exception thrown during on_init stage with message: %s \n", e.what());
    return CallbackReturn::ERROR;
  }
  return CallbackReturn::SUCCESS;
}
```

and could instead just do:
```c++
AdmittanceController::add_interfaces() {
  add_parameter(max_linear_velocity_, "Maximum linear velocity to clamp the twist command in meters per second");
  //...
  add_input<CartesianWrench>("ft_sensor");
  return CallbackReturn::SUCCESS;
}
```

**Comment by domire8** [comment]
I do like this idea! However, we've always made an effort to keep the terminology from ROS 2 / ROS 2 control for these callbacks and I'm not sure this deserves an exception. Do you have a strong opinion on this? If we want to do it, it would almost have to be part of modulo v5 to keep the breaking changes for the future as small as we can anticipate. I don't think it has to be done for this PR though.

**Comment by eeberhard** [comment]
Maybe I will take back the part about making `on_init() final`, and just `override` instead. But I do think adding a helpful virtual function like `add_interfaces()` is a good idea, just like `on_parameter_validate_callback()` is a custom AICA wrapper, and it makes it way simpler for derived controllers.

**Comment by domire8** [comment]
Issue created, will make it happen for v5

**Review by eeberhard — APPROVED** [review]
